segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,179 @@
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Manages the Write Queue Concurrency.
+ */
+@ThreadSafe
+class ConcurrencyManager {
+    //region Members
+
+    @VisibleForTesting
+    static final double SIGNIFICANT_DIFFERENCE = 0.1;
+    @VisibleForTesting
+    static final long MIN_FREQUENCY_MILLIS = 1000;
+    @VisibleForTesting
+    static final long STALE_MILLIS = MIN_FREQUENCY_MILLIS * 4;
+    private final int minParallelism;
+    private final int maxParallelism;
+    private final Supplier<Long> timeSupplier;
+    @GuardedBy("this")
+    private long recentTotalWrittenLength;
+    @GuardedBy("this")
+    private int recentWriteCount;
+    private final AtomicReference<Snapshot> lastSnapshot;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the ConcurrencyManager class.
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     */
+    ConcurrencyManager(int minParallelism, int maxParallelism) {
+        this(minParallelism, maxParallelism, System::nanoTime);
+    }
+
+    /**
+     * Creates a new instance of the ConcurrencyManager class.
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     * @param timeSupplier   A Supplier that returns the current time, in nanoseconds.
+     */
+    @VisibleForTesting
+    ConcurrencyManager(int minParallelism, int maxParallelism, Supplier<Long> timeSupplier) {
+        this.timeSupplier = Preconditions.checkNotNull(timeSupplier, "timeSupplier");
+        Preconditions.checkArgument(minParallelism <= maxParallelism, "minParallelism must be <= maxParallelism.");
+        this.minParallelism = minParallelism;
+        this.maxParallelism = maxParallelism;
+        this.recentTotalWrittenLength = 0;
+        this.recentWriteCount = 0;
+        this.lastSnapshot = new AtomicReference<>();
+        resetSnapshot();
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Records the fact that a write with the given length has just completed successfully.
+     *
+     * @param writeLength The length of the write that completed.
+     */
+    void writeCompleted(int writeLength) {
+        synchronized (this) {
+            this.recentWriteCount++;
+            this.recentTotalWrittenLength += writeLength;
+        }
+    }
+
+    /**
+     * Gets a value indicating the current (most recent) degree of parallelism.
+     *
+     * @return The result
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
```

Comment:
That's fine, but as it stand the logic of this method only works if it is invoked every MIN_FREQUENCY_MILLIS  to MIN_FREQUENCY_MILLIS *4 milliseconds, and if it is not, you get funky results. It would be preferable if it gave reasonable results/behaviour regardless of how often the caller invoked it.